### PR TITLE
[ADAG] Clean up channel read API

### PR DESCRIPTION
**Pull request by ruisearch42**
## Why are these changes needed?
After https://github.com/ray-project/ray/pull/45687 introduced `CompiledDAGRef` API, `end_read()` will be automatically called from MutableObjectManager destructor, and should be cleaned up.

This PR removes `begn_read()` and `end_read()` API and uses `read()` as a replacement for `begin_read()`.

This PR also gets and discards the execution result when `CompiledDAGRef.get()` is not called to avoid leak.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
As part of this PR could we do a couple other cleanups (sorry I missed these in #45687!):
- we need to release the value if ray.get() wasn't called by the user and the dag ref goes out of scope. otherwise, i think we will leak the value
- Could you clean up the `__str__` representation for CompiledDAGRef? These can be seen by the user and ideally we don't want to expose all of the private fields. It would be great if it could look somewhat similar to the normal ObjectRef API, so we include the hex ID and maybe the execution index, to indicate that it's different from normal ObjectRefs:
```python
CompiledDAGRef(00dc174c83599034ac4b146f35d6402c200cd6860100000002e1f505, execution_index=0)
```

File: python/ray/experimental/compiled_dag_ref.py [hunk 1]
```diff
@@ -37,10 +37,9 @@ def __init__(
 
     def __str__(self):
         return (
-            f"CompiledDAGRef(_dag={self._dag}, "
-            f"_execution_index={self._execution_index}, "
-            f"_ray_get_called={self._ray_get_called}, "
-            f"_dag_output_channels={self._dag_output_channels})"
+            f"CompiledDAGRef(dag_output_channels={self._dag_output_channels}, "
```
Comment (stephanie-wang) [inline python/ray/experimental/compiled_dag_ref.py:40]: Can we follow the ObjectRef syntax here? I.e. no `dag_output_channels` string and just print the IDs directly. The issue with `dag_output_channels` is that I don't think this means anything to the user (DAG users shouldn't need to know about the channel concept).

We should also remove `ray_get_called` - unlike execution_index, this field is not uniquely identifying so I don't think we should print it as part of __str__. You could move this to a `debug_info() -> str` method if you want.

Comment (ruisearch42) [reply python/ray/experimental/compiled_dag_ref.py:40]: Yeah let's discuss a bit more offline: we need to consider the special case of `CompositeChannel`, although it is not expected to be dag_output_channel.

**Review by stephanie-wang — APPROVED** [review]
Looks good! Could you also add the _dag_id to `CompiledDAG`'s __str__?

**Comment by ruisearch42** [comment]
> Looks good! Could you also add the _dag_id to `CompiledDAG`'s **str**?

Done.

**Pull request merged**